Upgrade to React 19

tyxla opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Contains work to add React 19 support.

Not intended to be merged directly, but rather to serve as a container for experimentation with all necessary updates.

See https://github.com/WordPress/gutenberg/issues/71336 for tracking the migration effort. 

## Why?
Keeping up to date with our latest framework dependencies and offering all the latest features.

## How?
Bumping deps and working on various fixes to ensure React 19 compat. Don't merge. 

## Testing Instructions
* Smoke test the app and look out for errors
* Check CI for errors.

### Testing Instructions for Keyboard
Same as above

## Screenshots or screencast <!-- if applicable -->
None

tyxla committed:
Bump to v19.2.4 (stable)

tyxla committed:
Bump to v19.2.4 (stable)

Mamaduka commented:
It's odd that we see some warnings in the unit test on this PR but not when I did the React 18.3 upgrade.

tyxla commented:
Indeed. Maybe they tinkered with some feature flags between 18.3 and 19 beta.

tyxla commented:
If I understand correctly, many of the test warnings seem to be because of how Ariakit retrieves the ref prop, it falls back to `element.ref`:

https://github.com/ariakit/ariakit/blob/73eca3904b97615a9fbc8c82eef7bc3afc86dae1/packages/ariakit-react-core/src/utils/misc.ts#L44

and that yields a warning in React 19. At the same time, I assume Ariakit will stay compatible with older React versions where `ref` was still handled specially. 

@diegohaz @DaniGuardiola any idea if we're planning to do anything about these warnings in Reakit?

DaniGuardiola commented:
@diegohaz could Ariakit possibly use some sort of React version detection instead and use it to determine what to do, rather than this fallback system that creates warnings?

diegohaz commented:
That part was updated to avoid those warnings (ref: https://x.com/diegohaz/status/1780794744534163888). But apparently it's not enough?

I still don't know the idiomatic way to fix that.

tyxla commented:
@diegohaz @DaniGuardiola I actually tried updating to the latest Ariakit version in this branch and it seems to resolve those warnings 👍 

Dani has started working on the update in #60992, and here we have one more reason to move it forward.

tyxla commented:
Updated to the RC. Hoping to have some time next week to play more thoroughly with it.

tyxla committed:
Update a snapshot test

tyxla committed:
Update a snapshot test

tyxla committed:
Fix context consumer serialization

tyxla committed:
Fix context consumer serialization

tyxla committed:
Fix some unit tests

tyxla committed:
Fix some unit tests

tyxla commented:
Fixed some tests, most of them were pretty straightforward. At first glance, the mobile tests seem to be failing because RN uses an old React version - we'll need a RN update to support React 19.

DaniGuardiola commented:
FYI, Frame Motion will be a blocker for now. More context: https://github.com/WordPress/gutenberg/issues/60975

Mamaduka commented:
React 19.0.0-rc.0.  is out.

> This is the exact build we'll release as 19.0, unless an issue is reported that requires a breaking change.

Source: https://x.com/acdlite/status/1797668537349328923

I also wonder if https://github.com/facebook/react/issues/29585 could be an issue for the block editor. Potentially, it can affect block re-ordering, but it will need some testing.

tyxla commented:
Updated to the latest RC. 

We'll still need to wait for RN support: 

https://github.com/callstack/react-native-testing-library/issues/1593

tyxla commented:
See https://github.com/callstack/react-native-testing-library/discussions/1698 for RN support for React 19 (hopefully H1 2025)

tyxla commented:
A few updates as I've been experimenting with making React 19 work in the meantime:

- Bumped to the latest React 19 RC in 63ac11d7c9153c74c67c393e8822235b76fe87fa
- Removed the now-unnecessary Ariakit patch in c557a6e342e9f7c736adbaf1f114003b224d00a5
- Removed the umd builds once more in 2d1ac297092fe14a435fc8ce0cc146ed8d35fe65
- Added a meta box container check to fix some errors in 39d6d4cb789eee72fff282c5663708b839bdbc0f
- Fixed `InserterBox` rendering in 3e8a81a53a56ea14a24ab65da42037ccd9e6a111
- Bumped `framer-motion` to the latest v12 alpha which introduces React 19 support in 6a9ada3dc414da51efb602be55224c6991580ab4

tyxla commented:
I'm experimenting with using `<link />` for loading separate component stylesheets on demand in https://github.com/WordPress/gutenberg/pull/66781

Mamaduka commented:
> I'm experimenting with using <link /> for loading separate component stylesheets on demand in https://github.com/WordPress/gutenberg/pull/66781

IIRC, we can also use suspense around components loading assets on demand and defer rendering. It could help render different previews.

Mamaduka commented:
The `react@19.0.0-rc.1` has been released - https://github.com/facebook/react/issues/29898#issuecomment-2477449973.

tyxla commented:
> The `react@19.0.0-rc.1` has been released - [facebook/react#29898 (comment)](https://github.com/facebook/react/issues/29898#issuecomment-2477449973).

Thanks for the heads up. A few updates:

- Bumped to the latest version in https://github.com/WordPress/gutenberg/pull/61521/commits/87bf9286249d17e995fa930413ada5d56a32d2a1
- Moved `framer-motion` dependency to the root to make it patchable in https://github.com/WordPress/gutenberg/pull/61521/commits/9a566e6d655e7637993de703f338c0995a06088a
- Patched `framer-motion@12.0.0-alpha.1` types to make type check pass in https://github.com/WordPress/gutenberg/pull/61521/commits/0ae8c137c61ae499e17355c16f2548787d465c42

Haven't done thorough testing, but it could be in a testable state at this point.

Note: React Native is still a blocker: https://github.com/callstack/react-native-testing-library/discussions/1698

tyxla committed:
Update useSuspenseSelect tests

tyxla committed:
Update useSuspenseSelect tests

Mamaduka commented:
React v19 is now stable - https://react.dev/blog/2024/12/05/react-19.

It seems that the biggest blocker here would be React Native.

tyxla commented:
> React v19 is now stable - https://react.dev/blog/2024/12/05/react-19.
> 
> It seems that the biggest blocker here would be React Native.

Thanks @Mamaduka - I've bumped this PR to the stable version.

I won't be able to work on it (test thoroughly and fix potential bugs) before next week, though.

Mamaduka commented:
Thanks, @tyxla!

I'll start looking into failing e2e tests if we still have some.

@youknowriad, [suggested](https://bsky.app/profile/riad.blog/post/3lclmhqr4xc2h) that we might try using separate web and mobile versions.

Shipping the new version in the Gutenberg plugin will help us gather early feedback on removed APIs.

tyxla commented:
> I'll start looking into failing e2e tests if we still have some.

Thanks so much for the help!

> @youknowriad, [suggested](https://bsky.app/profile/riad.blog/post/3lclmhqr4xc2h) that we might try using separate web and mobile versions.

Yeah, we'll likely have to do that - it's not clear if and when RN will be unblocked.

> Shipping the new version in the Gutenberg plugin will help us gather early feedback on removed APIs.

Absolutely.

Mamaduka commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -261,9 +261,6 @@ function Iframe( {
 		${ styles }
 		${ scripts }
 	</head>
-	<body>
-		<script>document.currentScript.parentElement.remove()</script>
-	</body>
```
@tyxla, what's the reason for removing this logic?

I think this is the cause of at least one e2e test failure. Ref #52405.

tyxla replied:
Back when I was testing with the beta, this was an essential fix to make the editor work (see https://github.com/WordPress/gutenberg/pull/61521/commits/7936e41e4058752881717b4a25826d3216a14cf2 for the full fix). Without it, the editor was crashing during its initial boot.

Mamaduka replied:
Thanks for the context, @tyxla!

I'll continue debugging. The `document.currentScript.parentElement.remove` removal is regressing #52405.

tyxla replied:
I appreciate your work on it, @Mamaduka! Feel free to revert it and seek a different fix. I've been delayed by other things, but hoping to get back to this one soon.

tyxla replied:
Reverted in https://github.com/WordPress/gutenberg/pull/61521/commits/3e1fbd41800090a38f2b94b51262d00d0d412e4d and did a second take in d4b4827eaff94858249805569b593f18228f0a55. I think we still need to omit the script removal part, because React will now ensure there's always one and only one `body` tag. The only other difference from the original code is that we're moving the frame loading just a bit later, after the other scripts.

ellatrix replied:
Could you comment on why it needs to be after the scripts now?

jsnajdr replied:
There must be an existing detailed GitHub discussion related to this. I remember investigating this issue sometime in 2024, and coming to some conclusions on how React portals work slightly differently in React 18 and 19.

Mamaduka replied:
Probably this thread https://github.com/WordPress/gutenberg/pull/52405#discussion_r1893056595.

jsnajdr replied:
Seems that the root cause of our `iframe` woes is this React PR that's part of v19: https://github.com/facebook/react/pull/25426

Here React implements first-class support for React owning the `html`, `head` and `body` elements. Respecting that they are singletons, i.e., that React should never remove them, but always only taking ownership of existing ones.

I'll need to figure out what this really means for our editor canvas iframe. The React v18 behavior was somewhat random and unreliable, like inserting a second `body` element, but by some accident and trial and error we made it work 🙂 Now we'll need to rethink the approach.

Mamaduka replied:
Random thought, I wonder if we can use that to our advantage and let React 19 handle our styles/scripts in the document head. Though I've not tested this with iframes.

Refs:
* https://react.dev/blog/2024/12/05/react-19#support-for-stylesheets
* https://react.dev/blog/2024/12/05/react-19#support-for-async-scripts
* https://react.dev/blog/2024/12/05/react-19#support-for-preloading-resources

tyxla replied:
Yeah, we should definitely try these out, they're quite ergonomic, and React handles a lot of the logic for deduplication, loading, unloading, and preloading. But I imagine this will be a separate effort after migrating to React 19. I did play with those early in the migration work btw, see #66781 (which is quite obsolete today, just look at the last 3 commits or so).

tyxla replied:
> I'll need to figure out what this really means for our editor canvas iframe. The React v18 behavior was somewhat random and unreliable, like inserting a second body element, but by some accident and trial and error we made it work 🙂 Now we'll need to rethink the approach.

That part looks pretty fragile, and I'm actually happy we have a chance to rearchitect it.

jsnajdr replied:
In 461825b I pushed a fix that is ugly, but seems to work. Let's see how many e2e tests become green now. We can then work on a better one.

Mamaduka commented:
* Update `framer-motion` to `v11.15.0`, which should be compatible with React 19 - 23bba7d.
* Removed `framer-motion` as a dependency from the root and reverted its patch.
* Resolved a couple of merge conflicts with package-lock.json and package.json.

Mamaduka commented:
I'm looking into failing e2e tests, and it seems the cause is a regression of #52405.

There's a behavior change in React 19 and `createPortal`.

**`createPortal` in React 18**

* The API rendered a second `<body>` tag and allowed our static one to work until the latter [removed itself](https://github.com/WordPress/gutenberg/blob/b226ceeeb96c66a17831c4dd5d5b7e8ec2d3e8ff/packages/block-editor/src/components/iframe/index.js#L265).
* Here's the CodeSandbox reproduction - https://codesandbox.io/p/sandbox/fnrn5n. You can see that both `body` elements are rendered inside the iframe.
* Hard to say if this is a correct behavior, but our hack from #52405 relied on it.

**`createPortal` in React 19**

* The API replaces the pre-rendered body element and regresses #52405.
* Here's the CodeSandbox reproduction - https://codesandbox.io/p/sandbox/iframe-react-19-c8m346.
* This seems like a correct behavior since a document can't have multiple body elements.

jsnajdr commented:
> I'm looking into failing e2e tests, and it seems the cause is a regression of https://github.com/WordPress/gutenberg/pull/52405.

@mamaduka The e2e failure looks like this:
```
      38 | 		await expect( canvasBody ).toHaveCSS( 'padding', '20px' );
    > 39 | 		await expect( canvasBody ).toHaveAttribute(
         | 		                           ^
      40 | 			'data-iframed-enqueue-block-assets-l10n',
      41 | 			'Iframed Enqueue Block Assets!'
      42 | 		);
```
What is interesting is that the `padding: 20px` test succeeded, the style added by the enqueued inline style is there. Also the other styles from `style.css` are present (`background-color: rgb(33, 117, 155)`). That means that at least some of the assets in `<head>` were loaded successfully. But the `<script>` assets failed to load. One reason for that could be that the iframe portal rendering removed the `<head>` element and aborted the loads that were inside it.

ellatrix commented:
Maybe in React 19, the new body element is no longer present at the time of the `load` event? Don't have much time to look into it, just guessing. Have you run the test locally and checked for errors in the iframe?

jmotes commented:
Hey everyone, is this PR the only progress toward supporting React 19? 

I'm especially interested in this since I lead development of a large Next.js based headless WordPress platform who's components we reference in our custom editor blocks. We are stuck using Next.js 14 / React 18 for that reason.

I'd be happy to help out however I can if anyone has a list of specific tasks that need done.

phil-the-dev commented:
@Mamaduka any updates on this? My team is blocked from upgrading to React 19 and Next 15 because of this dependency, and we can't be the only ones.

Mamaduka commented on package.json:
```diff
@@ -142,14 +142,14 @@
 		"puppeteer-core": "23.10.1",
 		"raw-loader": "4.0.2",
-		"react": "18.3.1",
+		"react": "19.0.0",
```
Do we plan to pin the versions to the latest React (19.1.1)? Ref: https://github.com/facebook/react/releases.

tyxla replied:
I would prefer we upgrade incrementally so have at least 1 release of the packages with 19.0, in case someone wants to stay on 19.0 and not upgrade to 19.1. Bumping to 19.1 will be the next thing after upgrading to v19.0 of course.

tyxla commented:
Won't have a chance to work on this anymore this week. Next up on my mind are:
- Editor iframe content doesn't always load
- Move away from deprecated umd builds; keep an eye on the react vendor chunks vs the element one
- We're still using v18 types
- Clean up the branch from the fixes and patches that we eventually reverted.

Mamaduka commented:
It seems the remaining e2e test failures are related to the `inert` attribute with Pattern Overrides.

Mamaduka committed:
Fix more 'inert' attributes

Mamaduka committed:
Fix more 'inert' attributes

Mamaduka commented:
@tyxla, I pushed some more fixes for the `inert` attribute.

I believe the mentioned error is related to `StrictMode` and the changes introduced to `useCallback` in React 19. The `useRefEffect` uses this hook internally with logic to mimic a cleanup method. I'll continue debugging this.

I cannot reproduce the error when running the app in production mode, using `SCRIPT_DEBUG: false` and a prod build.

P.S. I found a different error when `SCRIPT_DEBUG: false` and using the dev build for the plugin. This can be a common combo when developing or debugging the project.

## Screenshot
<img width="1146" height="199" alt="CleanShot 2025-09-02 at 08 25 48" src="https://github.com/user-attachments/assets/d5f03d14-7d36-42eb-a527-bed4c05653b8" />

tyxla commented:
Rebased, bumped all dependencies to the same version and migrated `JSX` manually to `React.JSX` (not particularly liking the result, I wish `JSX` was still available in the global namespace - worth trying some alternatives). 

Right now, the biggest blocker seems to be the UMD builds that we're relying on but are not available from React 19 on.

youknowriad commented:
Yes, we can build them ourselves, I actually did that in the past already but had to revert my PR until the React 19 upgrade (can 't recall the exact reason)

tyxla commented:
Makes sense - @jsnajdr @youknowriad, I'd appreciate some help with this part, it's a blocker for continuing the React 19 upgrade work.

jsnajdr committed:
Fix inert attributes

jsnajdr committed:
Fix inert attributes

jsnajdr committed:
Fix type errors

jsnajdr committed:
Fix type errors

jsnajdr committed:
Fix more type errors

jsnajdr committed:
Fix more type errors

jsnajdr committed:
Fix more types

jsnajdr committed:
Fix more types

jsnajdr committed:
Fix final batch of types

jsnajdr committed:
Fix final batch of types

jsnajdr committed:
Update docs

jsnajdr committed:
Update docs

jsnajdr committed:
Update changelogs

jsnajdr committed:
Update changelogs

Mamaduka commented:
I see that we're removing some public APIs from the `element` package: https://github.com/WordPress/gutenberg/pull/61521/changes#diff-106aefccde9db5c9cc1ab3dc05941ee7f155e429f6237a8c845d434b0b8ebb0e, which is likely a breaking change. Does anyone remember how we handled similar scenarios before?

ciampo commented:
@jsnajdr looking at the performance results from the CI tasks, some of the items have a considerable increase, although I guess it will need some investigation to understand whether it's something be concerned about.

Mamaduka commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -129,7 +129,14 @@ function Iframe( {
+			// If the `contentDocument` doesn't have a body yet, append a new one.
+			if ( ! doc.body ) {
+				const newBody = doc.createElement( 'body' );
+				newBody.className = 'temp';
```
Does this class serve any special purpose? Should we document it?

jsnajdr replied:
No, that's just a leftover from debugging. There were two `body` elements and I wanted to distinguish which one is which, so I added an unique classname. I'm lucky that I didn't use a worse name 🙂

Mamaduka reviewed (commented):
I can still reproduce the callback ref error originally reported by @tyxla!

## Steps to reproduce
1. Running in debug mode (haven't tested with prop yet).
2. TT5 theme.
3. Navigate to Site Editor > Patterns > Template Parts
4. Try changing to sorting.

I can reproduce consistently on this page.

## Screencast

https://github.com/user-attachments/assets/b2d16f0e-9fc4-4655-8a64-d467d1c5dca2

<img width="1338" height="474" alt="Image" src="https://github.com/user-attachments/assets/55bfc1fe-0bd7-4cea-8c06-86b2b8db0af4" />

tyxla commented:
> > how do you feel about landing this PR?
> 
> My plan is to:
> 
> 1. Land as much as possible as separate PRs, still on React 18. Most of the fixes are compatible.
> 2. Wait one week, until Feb 19, when the 7.0 beta 1 is branched and published.
> 3. Land the whole thing in `trunk` after that, so that it doesn't affect 7.0.
> 
> I'm not 100% if this is realistic, whether after Feb 19 the `trunk` branch will really be open for any post-7.0 work.

This plan sounds good and aligns with what we discussed earlier (also see [this comment](https://github.com/WordPress/gutenberg/issues/71336#issuecomment-3848070272)). AFAIK after the 7.0 beta 1 is published, any 7.0 work will be merged in the `wp/7.0` branch, so we should be alright to land this in `trunk`.

jsnajdr committed:
Remove MutableRefObject usages

jsnajdr committed:
Remove MutableRefObject usages

jsnajdr committed:
Fix user-event types

jsnajdr committed:
Fix user-event types

jsnajdr commented:
> looking at the performance results from the CI tasks, some of the items have a considerable increase, although I guess it will need some investigation to understand whether it's something be concerned about.

@ciampo Yes, this is true 🙁 Most notably, the `firstPaint` metric for `post-editor` is now 4x slower, it went up from 360ms to 1800ms. This result is consistent across many runs. And the `firstPaint` stat is something that is measured directly by the browser, reported as part of `performance.getEntriesByType( 'paint' )`.

I have no idea where this is coming from and whether it's real or imaginary. Something we need to investigate closer.

jsnajdr committed:
Add polyfill for findDOMNode

jsnajdr committed:
Add polyfill for findDOMNode

Mamaduka commented on packages/element/src/find-dom-node.ts:
```diff
@@ -0,0 +1,41 @@
+	}
+	const hostFiber = findHostFiber( fiber );
+	return hostFiber?.stateNode ?? null;
+}
```
Should we log a deprecation notice? We want to encourage folks to migrate away from deprecated APIs.

aduth replied:
I was thinking to comment something similar, and in general to have some idea of what it would take for us to feel comfortable actually removing it.

jsnajdr replied:
We could remove it in `trunk`, release a Gutenberg plugin without `findDOMNode`, and watch if there is any pushback.

I can add a deprecation notice. The React version has been logging a console warning for a few years now.

jsnajdr commented on packages/data/src/components/use-select/test/suspense.js:
```diff
@@ -229,7 +231,23 @@ describe( 'useSuspenseSelect', () => {
+		// - sibling prerendering that triggers an extra rerender after one component suspends;
+		// - fallback throttling that delays commit by 300ms and triggers an extra rerender.
+		expect( FastUI ).toHaveBeenCalledTimes( 4 );
+		expect( SlowUI ).toHaveBeenCalledTimes( 3 );
```
Here I added explanatory comments about the numbers of renders when using `useSuspenseSelect`. In the initial version of the PR @tyxla has been baffled by the reasons for this change, so I set out to explain it fully. There are two React 19 features that cause this:
- sibling prerendering, you might remember a little Twitter storm about it from time when React 19 beta came out (https://github.com/facebook/react/issues/29898#issuecomment-2477449973)
- fallback throttling (https://github.com/facebook/react/issues/31819)

Explaining this was one of my most rewarding Cursor/Claude sessions, where I opened the React codebase and started asking AI to explain why the code behaves in a certain way.

tyxla replied:
Nice, thanks for taking the time to go through it and for documenting it! 🙌

tyxla commented on package.json:
```diff
@@ -105,9 +105,9 @@
 		"prettier": "npm:wp-prettier@3.0.3",
 		"raw-loader": "4.0.2",
-		"react": "18.3.1",
+		"react": "19.2.4",
```
@jsnajdr I was previously planning to bump to React 19.0.x, to release at least one version of the packages with 19.0.x, so folks who can't upgrade to 19.2 for some reason (compat or any other - React 19.1 did introduce some breaking changes on its own), would have a way to at least upgrade to React 19. 

Did we drop that idea, or did it get lost in the huge pool of comments? [This comment](https://github.com/WordPress/gutenberg/pull/61521#discussion_r2309748563) was where I suggested this earlier, btw.

Curious to hear your thoughts in any case.

jsnajdr replied:
First of all, this PR started in May 2024, almost 2 years ago, so I wasn't always sure which of the comments and choices are still relevant today. For example, the Aug 2025 comment you're linking to, it was written at a time when 19.2 didn't exist yet.

From what I see in the [React changelog](https://github.com/facebook/react/blob/main/CHANGELOG.md), the really breaking changes, i.e., removals of APIs, are only in 19.0.0. Then 19.1 and 19.2 changelogs are long lists of bugfixes, improvements and new APIs, I don't see anything breaking there.

IMO it's better to do this kind of major upgrade just once, instead of delivering all the pain and uncertainty in several batches. "Check your plugin compat with WP 7.1" sounds like a better deal than "...and check it again in WP 7.2 and 7.3".

Are there any particular changes in 19.1 and 19.2 that stand out to you and are worth upgrading separately?

tyxla replied:
Hmm, I think I'm just overthinking it. I took another look at the changelog and I really don't see any concerns. Let's do it all in one chunk 👍

tyxla commented on packages/block-editor/src/components/block-list/block.js:
```diff
@@ -741,6 +741,31 @@ function BlockListBlockProvider( props ) {
 	);
 
+	// Use block visibility hook with data from existing useSelect to avoid extra subscription
+	const { isBlockCurrentlyHidden } = useBlockVisibility( {
```
Will this change also work with React 18? Should we move it to its own PR?

jsnajdr replied:
Yes, there are still opportunities to do standalone PRs, you flagged many of them. But it should be noted that the returns are diminishing. While the early spinoff PRs decreased the number of modified files from 240+ to 120, now we are talking about removing one or two files at a time.

tyxla replied:
Agreed, and I have no strong feelings either way, so I'll resolve my related comments.

tyxla commented on packages/block-editor/src/components/iframe/index.js:
```diff
@@ -169,7 +169,12 @@ function Iframe( {
+			// If the `contentDocument` doesn't have a body yet, append a new one.
+			if ( ! doc.body ) {
+				doc.documentElement.appendChild( doc.createElement( 'body' ) );
+			}
```
I really wish we could just rely on React doing its business with inserting when body when it should. How difficult would it be not to have to rely on the body to be there at all times?

jsnajdr replied:
Yes, the iframe portail rendering this is the most worrying part of the upgrade. The current iframe document does the following:
1. First there is a script in `head` that calls the `_load()` function, triggering a `setIframeDocument` call in the owning component, and starting the portal rendering.
2. Then, still in `head`, we load a lot of `style`s for the theme etc., asynchronously from remote URLs. The portal rendering already runs during this time.
3. Finally there is the `<body>` element, and it gets parsed and appeneded only after the styles have finished loading.

The result is that:
- we have to create our own `body` for portal rendering, because the one from the HTML parser doesn't exist yet.
- the styles are loaded only after then portal UI is rendered, often causing a flash of unstyled content in the editor.

It's very tricky to fix, especially before:
- delaying the portal render breaks may e2e tests. Because they often don't wait for the editor to appear, they rely on the accidental fact that it's always rendered and focused at the time when the relevant Playwright code runs.
- there's the quirky behavior when changing sort order of iframes, causing the iframe to partially unload and reload.

I already did one incremental improvement in #75619 and I plan to work on it further.

jsnajdr replied:
@tyxla @Mamaduka @aduth I finally achieved a little breakthrough in how to do the iframe/portal rendering properly, without any tricks. 🙂  The current solution:
- delays the `iframe` portal rendering until the `iframe` is loaded and all its document and window objects fully ready, using the standard `load` event.
- the `bodyRef` callback checks if the `body`'s document has a valid window (`defaultView`). If it doesn't, it returns early and doesn't call the other refs. This way, we don't need 5+ separate `defaultView` checks in relevant hooks (like `useWritingFlow`), the check is concentrated at one place. When React does its double-invoke disappear/reappear exercises, we handle them correctly.
- we no longer need to create our own `body` element and auto-remove it. The React portal will take ownership of the default `body` element.
- we also don't need a custom `_load` callback inside the iframe, the `load` event is fine.
- we make sure that the custom styles are loaded before we start rendering the blocks in a portal. That avoids a flash of unstyled content where the theme styles are applied late.

tyxla commented on packages/block-editor/src/components/observe-typing/index.js:
```diff
@@ -174,6 +171,7 @@ export function useTypingObserver() {
 				 * uncollapsed (shift) selection.
 				 */
 				function stopTypingOnSelectionUncollapse() {
+					const selection = defaultView.getSelection();
```
Can this change also go to its own PR?

tyxla commented on packages/block-editor/src/components/writing-flow/use-drag-selection.js:
```diff
@@ -41,6 +41,10 @@ export default function useDragSelection() {
 			const { defaultView } = ownerDocument;
 
+			if ( ! defaultView ) {
+				return;
```
Looks like these guards can be in their own PR. They should work just fine in React 18.

tyxla commented on packages/components/src/higher-order/with-notices/test/index.tsx:
```diff
@@ -80,16 +72,6 @@ const TestNoticeOperations = withNotices(
 	} )
 );
 
-describe( 'withNotices return type', () => {
```
Let's move these cleanups out in their own PR

tyxla commented on packages/components/src/number-control/index.tsx:
```diff
@@ -236,12 +236,8 @@ function UnforwardedNumberControl(
-				return stateReducerProp?.( baseState, action ) ?? baseState;
-			} }
+			value={ valueProp as string }
+			__unstableStateReducer={ numberControlStateReducer }
```
Nice cleanup, should it be moved out of the migration PR?

jsnajdr replied:
Moved out to #75822.

tyxla commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -93,7 +93,13 @@ function UnforwardedTooltip(
 	// See: https://github.com/WordPress/gutenberg/pull/65989
-	function addDescribedById( element: React.ReactElement ) {
+	function addDescribedById( el: React.ReactElement ) {
+		const element = el as React.ReactElement<
+			Pick<
+				React.HTMLAttributes< Element >,
+				'aria-describedby' | 'aria-label'
+			>
+		>;
```
Why was this assignment needed? Couldn't we just add the more detailed type to the attribute directly?

jsnajdr replied:
You mean changing the type of the `el` parameter? Then a similar assignment would just need to move to another place.

This is typical for elements that come from `children`. You can see a similar assignment in the `ResponsiveWrapper` component in `wp-components`. A component accepts only one child, and modifies it by adding some HTML attributes as props. `style`, `className`, `aria-something`. It is assumed that the child is a native HTML element or something compatible, but the `children` type doesn't say it. It's a generic `ReactElement`.

To make the types perfect, we'd have to modify the `children` type for these components. And I'm afraid that's a rabbit hole I'd never get out of 🙂

tyxla commented on packages/element/src/find-dom-node.ts:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Finds the DOM node of a React component instance.
+ *
+ * @deprecated since WordPress 6.2.0. Use DOM refs instead.
```
Did you mean 7.1?

tyxla commented on packages/element/src/find-dom-node.ts:
```diff
@@ -0,0 +1,49 @@
+	instance: React.ReactInstance
+): Element | Text | null {
+	deprecated( 'wp.element.findDOMNode', {
+		since: '6.2',
```
I guess you meant 7.1?

tyxla commented on packages/element/src/find-dom-node.ts:
```diff
@@ -0,0 +1,49 @@
+const HostText = 6;
+
+function findHostFiber( fiber: any ): any {
+	if ( fiber.tag === HostComponent || fiber.tag === HostText ) {
```
Are we intentionally keeping it simple and not checking if the component the node belongs to is mounted?

tyxla commented on packages/element/src/test/find-dom-node.js:
```diff
@@ -0,0 +1,36 @@
+import findDOMNodePolyfill from '../find-dom-node';
+
+describe( 'findDOMNode', () => {
+	it( 'should find DOM node rendered by a nested child component', () => {
```
Can we add a few more basic tests? I'd cover at least the basic scenario `Returns `null` if passed `null` or any random object`. More inspiration can be found in React's tests:

https://github.com/facebook/react/blob/a7c898d83a991c48f3981fcc65d969f1d90d80a1/packages/react-dom/src/__tests__/findDOMNodeFB-test.js

Maybe we don't need to get too complex, but something to keep in mind as we seek the right balance in the polyfill.

jsnajdr replied:
I updated the `findDOMNode` polyfill and now it's more reliable and resilient than before, I'd say it's on par with the original implementation 🙂 There are also more tests, inspired by the React impl.

> Are we intentionally keeping it simple and not checking if the component the node belongs to is mounted?

This is something that was OK even in the original simplified implementation. Unmounted component doesn't have the `_reactInternals` field with the fiber pointer. `findDOMNode` returns `null` in that case.

> I guess you meant 7.1?

Also updated. It's a bummer that we forgot to add the deprecation notice back in 6.2, together with `render`, `hydrate` and `unmountComponentAtNode`.

tyxla commented on packages/element/src/react-platform.ts:
```diff
@@ -22,36 +15,13 @@ import { createRoot, hydrateRoot } from 'react-dom/client';
- * @deprecated since WordPress 6.2.0. Use `createRoot` instead.
- * @see https://react.dev/reference/react-dom/render
- */
-export { render };
```
There are a few things going on here:

* Removing `findDOMNode`
* Removing `render` in favor of `createRoot`
* Removing `hydrate` in favor of `hydrateRoot`
* Removing `unmountComponentAtNode` in favor of `root.unmount`

Let's make sure to document these breaking changes in the CHANGELOG of the element package. These will also need to be covered in any dev notes and make.wordpress.org/core posts.

jsnajdr replied:
The `element` changelog is now updated with a "breaking changes" section that will trigger a major version bump of the released package.

tyxla commented on packages/element/package.json:
```diff
@@ -45,13 +45,14 @@
-		"@types/react-dom": "^18.3.1",
+		"@types/react": "^19.2.13",
+		"@types/react-dom": "^19.2.3",
+		"@wordpress/deprecated": "file:../deprecated",
```
Any bundle size impact by including `@wordpress/deprecated` as a dependency? I think it only depends on `@wordpress/hooks` which has no dependencies itself, but still, worth checking.

jsnajdr replied:
Maybe in some very isolated and specialized use case where we use `wp-element` and nothing else, adding `wp-deprecated` might introduce a new dependency. But in just about any reasonable WordPress environment I can imagine, the entire Gutenberg or its big part is loaded anyway, and `wp-deprecated` is already used by many other components.

One environment where it might matter is frontend widgets that use Interactivity API. But Interactivity API uses preact and doesn't use `wp-element` at all.

tyxla commented on packages/element/README.md:
```diff
@@ -151,11 +151,17 @@ _Changelog_
-Finds the dom node of a React component.
+> **Deprecated** since WordPress 6.2.0. Use DOM refs instead.
+
+Finds the DOM node of a React component instance.
```
Let's also highlight that this is a polyfill and lacks some of the intricacies of the original React utility, like error handling or handling special cases during strict mode.

jsnajdr replied:
I think that at this moment the polyfill is fully on par with the original implementation. The only thing missing is React's own deprecation warnings. I guess we have enough of our own WordPress ones 🙂

tyxla commented on packages/element/src/find-dom-node.ts:
```diff
@@ -0,0 +1,49 @@
+	instance: React.ReactInstance
+): Element | Text | null {
+	deprecated( 'wp.element.findDOMNode', {
+		since: '6.2',
```
Should we also aim to completely remove it at some point?

jsnajdr replied:
I personally would be happiest if we could just remove `findDOMNode` together with `hydrate`, `render` and `unmountComponentAtNode`, announce a compat break early and let everyone migrate.

> Should we also aim to completely remove it at some point?

The question is: at what point? If we for some reason are unwilling to remove it right now, what will change in the future that will make the removal a better idea? I'm afraid that nothing will change. There will always be plugins that are legacy, unmaintained, but still widely used and useful, that use legacy class components and `findDOMNode`.

Mamaduka replied:
> I'm afraid that nothing will change.

An unfortunate truth :(

tyxla commented on test/e2e/specs/site-editor/patterns.spec.js:
```diff
@@ -189,6 +189,63 @@ test.describe( 'Patterns', () => {
 		await expect( patterns.item ).toContainText( 'Unsynced footer' );
 	} );
+
+	test( 'sort patterns', async ( {
```
Does this e2e test belong to its own PR?

jsnajdr replied:
Yes, spinned off in #75823, already merged :+1:

tyxla reviewed (commented):
Exciting to see this close to being ready, kudos for the great work on it folks 🙌 

Left a few thoughts as I was reviewing the changes. I think some of them can be moved out

My only concern is that we're immediately bumping to the latest React version, and by that we're not giving a chance to folks to perform a more granular migration (18.3 to 19.0 to 19.1 or 19.2).

jsnajdr committed:
Try to do iframe loading properly

jsnajdr committed:
Try to do iframe loading properly

Mamaduka commented on test/emptytheme/templates/index.html:
These changes could also be a good candidate for a separate PR.

jsnajdr replied:
Yes, that's a quite important standalone PR. Here it is: #75894.

Mamaduka commented on packages/e2e-tests/plugins/iframed-enqueue-block-assets/script.js:
```diff
@@ -1,4 +1,4 @@
 window.addEventListener( 'load', () => {
-	document.body.dataset.iframedEnqueueBlockAssetsL10n =
+	document.documentElement.dataset.iframedEnqueueBlockAssetsL10n =
```
This changes the nature of the test. If core tests can't correctly add and test attributes for the iframe body, the same will be true for any custom plugin.

I think there will be enough cases where consumers inject custom attributes or classes to iframe body. Related - #17854.

jsnajdr replied:
The trouble is that the `body` element is fully owned by React. The portal rendering code in the `Iframe` component says:
```js
createPortal( <body ref className></body> )
```
And this React render now removes the `data-*` attributes that are not in the `<body>` JSX markup. That's why the test starts failing.

My question is (I don't know the answer yet): did this plugin that adds something custom to `<body>`, did it ever work as intended? The previous logic is that there are two `<body>` elements. One added by the HTML parser, one added by our custom rendering logic. And after a short while one of them is removed.

Did the plugin add the `data-*` attribute to the `<body>` element that was removed, or to the element that stayed alive? In other words, was the `data-*` attribute added permanently? Or did the test happen to catch the very short period of time when it was present?

> Related - https://github.com/WordPress/gutenberg/issues/17854.

Yes, that looks like a reasonable request, but the solution probably isn't adding the `className` using DOM API to React-owned elements. React is going to overwrite these changes.

Mamaduka replied:
I don't think this should be a blocker for the upgrade; it's just something we should definitely track and provide an alternative for. Our previous hacks only worked because of unexpected side effects in the React code.

Also agreed on https://github.com/WordPress/gutenberg/issues/17854. There could be a better API that allows consumers to set their own classes on the iframe body.

jsnajdr replied:
Today I tried one idea how to fix this issue: use `flushSync` to do the initial render and `body` hydration as early as possible:
```js
function onLoad() {
  flushSync( () => setIframeDocument( iframeDocument ) );
}
```
That was supposed to solve the timing issues, but it didn't work, so I reverted it again.

> There could be a better API that allows consumers to set their own classes on the iframe body.

Maybe it's already possible to do it right, and even used somewhere. The key is to use React to add the `body` attributes. Just take this ref callback and somehow attach it to just about any element inside the `iframe`:
```js
const ref = useRefEffect( ( node ) => {
  node.ownerDocument.body.dataset.foo = 'bar';
  return () => {
    delete node.ownerDocument.body.dataset.foo;
  }
} );

return <div ref={ ref } />;
```
Can this be done by hooking into one of existing Gutenberg filters? Something like `editor.BlockEdit`, but global instead of per-block? It's quite likely that it already exists.

Mamaduka replied:
I think it depends on the case. Some values can be exposed via the REST API, as with `post_class`, and we could then copy them.

As far as I know, we're trying to avoid introducing filters like `BlockEdit`. They can be perf bottlenecks, preventing us from making optimizations without breaking backward compatibility. The iframe component is used in many places, including block/pattern/template previews, the editor canvas, etc. Global filters could affect all of them.

Maybe there's a new take on extensibility and similar filters, but I've not read it anywhere. cc @youknowriad

jsnajdr committed:
findDOMNode: improve tests and implementation

jsnajdr committed:
findDOMNode: improve tests and implementation

jsnajdr committed:
Update element changelog with breaking changes

jsnajdr committed:
Update element changelog with breaking changes

tyxla commented:
On the inserter metric changes: I'm assuming this has something to do with the changes to iframe portal rendering that are part of the upgrade. Previously, we were relying on hacks and were manually constructing some of the document, and now with React handling some of that (and us waiting for it accordingly), changes to metrics are expected. Worth verifying, of course.

jsnajdr committed:
Element: start exporting new React APIs

jsnajdr committed:
Element: start exporting new React APIs

jsnajdr committed:
Fix admin-ui tests: React 19 doesn't log errors inside act

jsnajdr committed:
Fix admin-ui tests: React 19 doesn't log errors inside act

jsnajdr committed:
Fix inert attribute check in e2e test

jsnajdr committed:
Fix inert attribute check in e2e test

jsnajdr committed:
Fix props spread

jsnajdr committed:
Fix props spread

tyxla committed:
Align grid + dashboard widgets with React 19

packages/grid kept React 18 in its peerDependencies, which forced npm to
install a nested React 18 inside widgets/activity and widgets/welcome
even though the rest of the tree is on 19.2.4. Bump grid's peerDeps and
drop the unnecessary direct `react` dep from the two widgets — other
widgets already rely on the top-level React.

Also restores the nested React 18.3.1 install for test/native (still
pinned to 18 for React Native compatibility) that was missing from the
rebase's auto-merged lockfile.

tyxla committed:
Dashboard: fix inert attribute for React 19

React 19 treats `inert` as a boolean attribute, not a string. The
widget-chrome and widget-picker components were passing `inert: ''`,
which React 19 rejects with a console.error. Replace with the existing
branch pattern: `inert={ condition || undefined }` so the attribute is
omitted entirely when not active.

tyxla committed:
Pass undefined to useRef calls without an initial value

React 19 removed the no-argument overload of useRef. Each affected call
now passes \`undefined\` explicitly, widening the ref type to include
\`undefined\` where it wasn't already. Touches new media-editor, grid,
and keyboard-shortcuts code that landed on trunk after the React 19
branch was originally written.

tyxla committed:
Import JSX from 'react' instead of relying on a global

React 19 no longer declares JSX as a global namespace. The
media-editor module's component-prop type definitions used \`JSX.Element\`
directly; import \`JSX\` alongside the other named types from 'react'.

tyxla committed:
Grid: fix inert attribute for React 19

React 19 expects \`inert\` to be a boolean, not a string. The
dashboard-grid and dashboard-lanes item components were passing
\`{ inert: '' }\` via spread; replace with \`inert={ dragging || undefined }\`
matching the branch's established pattern (omit the attribute when
inactive rather than emit \`inert="false"\`).

tyxla committed:
Tighten cloneElement type in renderSlotWithChildren

React 19's cloneElement second-arg type requires a type-compatible
props shape. Cast the slot to \`ReactElement<{ children?: ReactNode }>\`
so the children injection type-checks (matches the function's
documented intent of overwriting any subtree passed on the slot).

tyxla committed:
Grid: type the actionableArea child props in dashboard-grid

React 19's stricter \`React.ReactElement\` typing leaves \`child.props\`
as \`unknown\`, so destructuring or passing \`actionableArea\` through
\`cloneElement\` fails. Cast the child to
\`React.ReactElement<{ actionableArea?: React.ReactNode }>\` so the
lift-and-strip pattern in DashboardGrid type-checks.

tyxla committed:
Theme, UI: changelog entries for the React 19 bump

Both packages bumped their React peerDependencies from ^18 to ^19.2.4
as part of the broader React 19 upgrade; add the corresponding
"Upgrade React to v19" entry under "Unreleased" so the per-package
changelog checks pass.

tyxla committed:
RTC: Re-render collaborators overlay when the block tree changes

A remote block insert can land in the DOM after the inserting user's
cursor awareness arrives. Without a block-tree dependency, the overlay
queries [data-block] once, finds nothing, and never retries — leaving
the remote cursor invisible until the collaborator moves again.

jsnajdr reviewed (approved):
This is now ready to merge, let's go 🤞
